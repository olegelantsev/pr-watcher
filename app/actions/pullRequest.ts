import fetch from 'cross-fetch';
import {
  GetState,
  Dispatch,
  PullRequest,
  User,
  Reviewer,
  AccountSetting,
  REQUEST_PULL_REQUESTS,
  RECEIVE_PULL_REQUESTS,
  ReceivePullRequestsAction
} from '../reducers/types';

// https://docs.atlassian.com/bitbucket-server/rest/7.2.3/bitbucket-rest.html#idp281

export function requestPullRequests() {
  return {
    type: REQUEST_PULL_REQUESTS
  };
}

export async function fetchRepoSlags(accountSetting: AccountSetting) {
  const response = await fetch(
    `${accountSetting.url}/rest/api/1.0/projects/${accountSetting.projectSlug}/repos`,
    {
      method: 'get',
      headers: new Headers({
        Authorization: `Bearer ${accountSetting.token}`
      })
    }
  );
  if (response.status >= 400) {
    throw new Error(`Server responded with ${response.status}`);
  }
  const data = await response.json();
  return data.values.map((x: any) => x.slug);
}

function convertUser(x: any) {
  return {
    displayName: x.displayName,
    email: x.emailAddress,
    id: x.id,
    name: x.name
  } as User;
}

function parseReviewers(reviewers: [any]) {
  return reviewers.map((reviewer: any) => {
    return {
      approved: reviewer.approved,
      lastReviewedCommit: reviewer.lastReviewedCommit,
      user: convertUser(reviewer.user)
    } as Reviewer;
  });
}

// function resetNewPrCount() {
//   return {
//     type: RESET_NEW_PR_COUNT
//   };
// }

function receivePullRequests(
  pullRequests: Array<PullRequest>,
  error?: string
): ReceivePullRequestsAction {
  return {
    type: RECEIVE_PULL_REQUESTS,
    pullRequests,
    receivedAt: Date.now(),
    error,
    status: error != null ? 'error' : 'success'
  };
}

export async function fetchPullsRequests(
  baseUrl: string,
  projectKey: string,
  repositorySlug: string
) {
  const response = await fetch(
    `${baseUrl}/rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/pull-requests`
  );
  const data = await response.json();
  return data.values.map((x: any) => {
    return {
      title: x.title,
      description: x.description,
      createdDate: x.createdDate,
      updatedDate: x.updatedDate,
      author: convertUser(x.author.user),
      reviewers: parseReviewers(x.reviewers),
      link: x.links.self[0].href,
      state: x.state
    } as PullRequest;
  });
}

async function fetchPullRequestsAndMerge(
  repoSlugs: Array<string>,
  accountSetting: AccountSetting
): Promise<Array<PullRequest>> {
  const listOfPullRequests = await Promise.all(
    repoSlugs.map((repoSlug: string) =>
      fetchPullsRequests(
        accountSetting.url,
        accountSetting.projectSlug,
        repoSlug
      )
    )
  );
  const merged = [].concat(...listOfPullRequests);

  return merged;
}

export function fetchAll() {
  return (dispatch: Dispatch, getState: GetState) => {
    const { accountSetting } = getState();

    if (accountSetting == null) {
      throw Error("Can't fetch if account isn't added.");
    }

    dispatch(requestPullRequests());

    return fetchRepoSlags(accountSetting)
      .then(repoSlugs => {
        return fetchPullRequestsAndMerge(repoSlugs, accountSetting);
      })
      .then((pullRequests: Array<PullRequest>) => {
        return dispatch(receivePullRequests(pullRequests));
      })
      .catch(err => {
        dispatch(receivePullRequests([], String(err)));
      });
  };
}
