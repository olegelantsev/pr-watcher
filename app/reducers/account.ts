import { UPDATE_ACCOUNT, AccountSetting, UpdateAccountSetting } from './types';

const initialState = {
  projectSlug: 'PRJ',
  token: '123',
  url: 'http://localhost:5000'
} as AccountSetting;

export default function account(
  state: AccountSetting = initialState,
  action: UpdateAccountSetting
): AccountSetting {
  switch (action.type) {
    case UPDATE_ACCOUNT:
      return {
        ...state,
        projectSlug: action.accountSetting.projectSlug,
        token: action.accountSetting.projectSlug,
        url: action.accountSetting.url
      };
    default:
      return state;
  }
}
