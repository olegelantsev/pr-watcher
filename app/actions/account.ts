import {
  AccountCheckError,
  AccountSetting,
  UPDATE_ACCOUNT,
  UpdateAccountSetting,
  errorType,
  AccountCheckProgress,
  ACCOUNT_CHECK_IN_PROGRESS,
  Dispatch,
  ACCOUNT_CHECK_ERROR
} from '../reducers/types';
import { fetchRepoSlags } from './pullRequest';

export function updateAccount(
  accountSetting: AccountSetting,
  checking: boolean,
  error?: errorType
): UpdateAccountSetting {
  return {
    type: UPDATE_ACCOUNT,
    accountSetting,
    checking,
    error
  };
}

function accountCheckInProgress(): AccountCheckProgress {
  return {
    type: ACCOUNT_CHECK_IN_PROGRESS
  };
}

function accountCheckError(
  status: errorType,
  error: string
): AccountCheckError {
  return {
    type: ACCOUNT_CHECK_ERROR,
    error,
    status
  };
}

async function checkAccount(accountSetting: AccountSetting): Promise<boolean> {
  await fetchRepoSlags(accountSetting);
  return true;
}

export function checkAndUpdateAccount(accountSetting: AccountSetting) {
  return (dispatch: Dispatch) => {
    dispatch(accountCheckInProgress());

    return checkAccount(accountSetting)
      .then(() => {
        return dispatch(updateAccount(accountSetting, false, undefined));
      })
      .catch(err => {
        dispatch(accountCheckError('error', String(err)));
      });
  };
}
