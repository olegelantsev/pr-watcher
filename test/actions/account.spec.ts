import { spy } from 'sinon';
import nock from 'nock';
import * as actions from '../../app/actions/account';
import * as repoResponse from './stubs/repos.json';
import * as pullRequestsResponse from './stubs/pullRequests.json';
import * as twoReposResponse from './stubs/repos2.json';
import * as repo2PullRequestResponse from './stubs/repo2_pullRequests.json';
import {
  REQUEST_PULL_REQUESTS,
  ApplicationState,
  AccountSetting,
  ACCOUNT_CHECK_IN_PROGRESS
} from '../../app/reducers/types';

describe('account actions', () => {
  it('checks account successfully', async () => {
    nock('http://bitbucket.com')
      .get('/rest/api/1.0/projects/123/repos')
      .reply(200, repoResponse, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      });
    return new Promise(resolve => {
      Date.now = jest.fn(() => 1359075920);
      const fn = actions.checkAndUpdateAccount({
        url: 'http://bitbucket.com',
        projectSlug: '123',
        token: 'some-token'
      } as AccountSetting);
      const dispatch = spy();
      const getState = () =>
        ({
          accountSetting: {
            checking: false,
            error: undefined
          }
        } as ApplicationState);
      fn(dispatch, getState);
      setTimeout(() => {
        const calls = dispatch.getCalls();
        expect(calls.length).toBe(2);

        expect(dispatch.getCall(0).args[0]).toEqual({
          type: ACCOUNT_CHECK_IN_PROGRESS
        });

        expect(dispatch.getCall(1).args[0]).toMatchSnapshot();
        resolve();
      }, 50);
    });
  });

  it('checks account with error', async () => {
    nock('http://bitbucket.com')
      .get('/rest/api/1.0/projects/123/repos')
      .reply(404, repoResponse, {
        'Access-Control-Allow-Origin': '*'
      });
    return new Promise(resolve => {
      Date.now = jest.fn(() => 1359075920);
      const fn = actions.checkAndUpdateAccount({
        url: 'http://bitbucket.com',
        projectSlug: '123',
        token: 'some-token'
      } as AccountSetting);
      const dispatch = spy();
      const getState = () =>
        ({
          accountSetting: {
            checking: false,
            error: undefined
          }
        } as ApplicationState);
      fn(dispatch, getState);
      setTimeout(() => {
        const calls = dispatch.getCalls();
        expect(calls.length).toBe(2);

        expect(dispatch.getCall(0).args[0]).toEqual({
          type: ACCOUNT_CHECK_IN_PROGRESS
        });

        expect(dispatch.getCall(1).args[0]).toMatchSnapshot();
        resolve();
      }, 150);
    });
  });
});
