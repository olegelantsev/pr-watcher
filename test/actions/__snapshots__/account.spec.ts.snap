// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account actions checks account successfully 1`] = `
Object {
  "accountSetting": Object {
    "projectSlug": "123",
    "token": "some-token",
    "url": "http://bitbucket.com",
  },
  "checking": false,
  "error": undefined,
  "type": "UPDATE_ACCOUNT",
}
`;

exports[`account actions checks account with error 1`] = `
Object {
  "error": "Error: Server responded with 404",
  "status": "error",
  "type": "ACCOUNT_CHECK_ERROR",
}
`;
